Adds a fully functional engine + completes code of incomplete classes.

L-Orsi opened the pull request:
Adds game engine + completes unfinished classes needed for it to run.

- Tests are to be added in a separate PR

ernestmc commented on src/bug.cpp:
```diff
@@ -74,11 +105,18 @@ namespace simulation
     }
-    std::shared_ptr<Bug> newborn_bug = std::make_shared<Bug>(free_adjacent);
+    breed_counter_ = 0;
+    std::shared_ptr<Bug> newborn_bug = std::make_shared<Bug>(free_adjacent, !is_prey_, rounds_to_breed_, rounds_to_starve_, name_ + ".");
```
nice surname "." :rofl:

L-Orsi replied:
Hahah tbh, It was useful to both occupy little space on the screen and still know who the parent the new bug was.

ernestmc commented on src/bug_notifier.cpp:
```diff
@@ -9,22 +9,22 @@ namespace simulation
   BugNotifier::BugNotifier() {}
 
-  void BugNotifier::add_observer(std::shared_ptr<BugObserver> observer) {
+  void BugNotifier::add_observer(BugObserver* observer) {
```
Why move to raw pointer?

L-Orsi replied:
Explained in reply of comment above.

ernestmc commented on src/bug_notifier.cpp:
```diff
@@ -9,22 +9,22 @@ namespace simulation
 
   void BugNotifier::notify_death(std::shared_ptr<Bug> dead_bug) {
-    for(const auto observer : observers_) {
+    for(auto observer : observers_) {
```
```suggestion
    for(const auto& observer : observers_) {
```

ernestmc commented on src/map_cell.cpp:
```diff
@@ -39,18 +39,25 @@ void MapCell::move_bug(const std::shared_ptr<MapCell> destination) {
 }
 
-const std::array<const std::shared_ptr<MapCell>, ADJACENT_CELLS>& MapCell::get_adjacents() const {
+std::array<std::shared_ptr<MapCell>, ADJACENT_CELLS>& MapCell::get_adjacents() {
```
Why is this method not const anymore?

ernestmc commented on src/simulation_engine.cpp:
```diff
@@ -1,21 +1,92 @@
 
-void SimulationEngine::populate_map_with_species() {}
+void SimulationEngine::start_simulation(uint16_t n_rounds) {
+  for(;n_rounds && simulate_round();n_rounds--);
```
Submit code [here](http://www.ioccc.org/) :laughing:

L-Orsi replied:
I like that line too much, will add a comment for clarity but I want to keep it hahah

ernestmc commented on src/simulation_engine.cpp:
```diff
@@ -1,21 +1,92 @@
+  uint16_t species_left = 0;
+  std::cout << "--------" << std::endl;
+  std::cout << "[SimulationEngine] New Round" << std::endl;
+    map_->print_map();
```
indentation

ernestmc commented on src/simulation_engine.cpp:
```diff
@@ -1,21 +1,92 @@
+  // Iterates over cells and makes every bug move/breed/eat/starve accordingly.
+  for(auto cell: map_->get_cells()) {
+    auto bug_ref = cell->get_bug();
+    if(bug_ref && !bug_ref->has_played()) {
```
Why do we need to check if it has already played? Can we iterate over the same bug more than once?

L-Orsi replied:
If we don't, as we're iterating over cells and not over bugs, if a bug moves to a cell that was not yet iterated over, the bug would have two turns in he same round.

ernestmc replied:
got it

ernestmc commented on src/simulation_engine.cpp:
```diff
@@ -1,21 +1,92 @@
+    if(cell->get_bug()) {
+      species_left++;
+      only_doo &= !cell->get_bug()->is_prey();
+      only_ants &= cell->get_bug()->is_prey();
```
Again, I don't like checking if it's a prey to see if it's an ant. You should had been more explicit

L-Orsi replied:
Oh, I thought we were aware that this was a weak spot of the code if we eventually wanted to scale it, but didn't know it was a requirement to change it. 
If that's the case, I'd do so in a separate PR though, It's a mid-sized change, I don't want to increase the size of this PR that is already too big.

ernestmc replied:
not required to change it. I was aware that we agreed to leave it like. Just stressing it again because I'm tough :stuck_out_tongue_closed_eyes:

ernestmc commented on src/simulation_engine.cpp:
```diff
@@ -1,21 +1,92 @@
+bool SimulationEngine::simulate_round() {
+  uint16_t species_left = 0;
+  std::cout << "--------" << std::endl;
+  std::cout << "[SimulationEngine] New Round" << std::endl;
```
It would be cool to print the round number.

ernestmc commented on src/simulation_engine.cpp:
```diff
@@ -1,21 +1,92 @@
+    for(auto j=0; j<DOODLEBUGS; j++) {
+      (*cells_it)->set_bug(std::make_shared<Bug>(*cells_it, BugDescriptorDatabase::get_doodlebug_descriptor()));
+      (*cells_it)->get_bug()->add_observer(this);
+      cells_it++;
```
what happens if you reach the end of iterator?

ernestmc replied:
Beware that the `end` iterator points _past _ the last element. See [here](https://en.cppreference.com/w/cpp/container/vector/end)
So if you reach this line you already got a null pointer exception

ernestmc reviewed (commented):
Left some comments

L-Orsi commented on include/simulator/bug.hpp:
```diff
@@ -35,16 +42,38 @@ class Bug : public BugNotifier {
+  std::string get_name() const;
+
+  /** Gets the cell where the bug is at. */
+  std::shared_ptr<MapCell> get_cell() {
```
+1, and the definition shouldn't be part of the header.

L-Orsi commented on include/simulator/bug_notifier.hpp:
```diff
@@ -17,7 +17,7 @@ class Bug;  // Forward declaration.
 class BugNotifier {
  private:
-  std::set<std::shared_ptr<BugObserver>> observers_;
+  std::unordered_set<BugObserver*> observers_;
```
Because of the issue talked offline about creating a `shared_ptr` to `this`. 
I know we can kind of trick this by using the tools we discussed offline as well, however, I don't thing it's a good practice to use so here. Usually, the creation of the content held by a `shared_ptr` represents ownership, which is not the case here (as the class does not own itself). By doing some research I found out that, even in cpp, the usual way to code this pattern is using raw pointers. I don't want to get to philosophical here, but the issue with raw pointers are related to memory allocation, and here we're not creating and/or destructing anything, so it's the perfect use case for raw pointers to me.

L-Orsi commented on include/simulator/simulation_engine.hpp:
```diff
@@ -9,22 +9,38 @@
 
-constexpr uint8_t ANTS_INIT_NUMBER = 10;
-constexpr uint8_t DOODLEBUGS_INIT_NUMBER = 5;
+constexpr uint8_t ANTS_INIT_NUMBER = 5;
+constexpr uint8_t DOODLEBUGS_INIT_NUMBER = 2;
```
Sounds like a feature. Maybe it can be added in a future PR alongside the is_prey refactor.

L-Orsi commented:
@ernestmc I addressed the comments, besides two of them that maybe are worth a new PR?

- Changing the prey idea to ants/doodlebugs
- Input number of bugs through command line and/or file

I wonder if we'll have time go get those two features in.

ernestmc commented:
The suggested features do not need to be implemented. They are for future improvement.
